File: Cpp/src/objects/output/Statistics.cpp
#include "Statistics.h"

Statistics::Statistics() {
    this->configurations = new vector<StatConfigurations*>();
}
Statistics::~Statistics() {
    delete this->configurations;
}

void Statistics::Serialize(Json::Value& root) {
    root["name"] = name;
    root["status"] = status;
    root["overall_texts"] = overall_texts;
    root["overall_batches"] = overall_batches;
    root["overall_calls"] = overall_calls;
    root["calls_system"] = calls_system;
    root["calls_data"] = calls_data;
    root["overall_docs"] = overall_docs;
    root["docs_processed"] = docs_processed;
    root["docs_failed"] = docs_failed;
    root["docs_responded"] = docs_responded;
    root["overall_colls"] = overall_colls;
    root["colls_processed"] = colls_processed;
    root["colls_failed"] = colls_failed;
    root["colls_responded"] = colls_responded;
    root["colls_documents"] = colls_documents;
    root["latest_used_app"] = latest_used_app;
    root["used_apps"] = used_apps;
    
	for (int i = 0; i < configurations->size(); ++i )  {
        Json::Value value;
        configurations->at(i)->Serialize(value);
        root["configurations"].append(value);
	}
}

void Statistics::Deserialize(Json::Value& root) {
    name = root.get("name", "").asString();
    status = root.get("status", "").asString();
    overall_texts = root.get("overall_texts", 0).asInt();
    overall_batches = root.get("overall_batches", 0).asInt();
    overall_calls = root.get("overall_calls", 0).asInt();
    calls_system = root.get("calls_system", 0).asInt();
    calls_data = root.get("calls_data", 0).asInt();
    overall_docs = root.get("overall_docs", 0).asInt();
    docs_processed = root.get("docs_processed", 0).asInt();
    docs_failed = root.get("docs_failed", 0).asInt();
    docs_responded = root.get("docs_responded", 0).asInt();
    overall_colls = root.get("overall_colls", 0).asInt();
    colls_processed = root.get("colls_processed", 0).asInt();
    colls_failed = root.get("colls_failed", 0).asInt();
    colls_responded = root.get("colls_responded", 0).asInt();
    colls_documents = root.get("colls_documents", 0).asInt();
    latest_used_app = root.get("latest_used_app", "").asString();
    used_apps = root.get("used_apps", "").asString();

    if (NULL == this->configurations) {
        this->configurations = new vector<StatConfigurations*>();
    }
    
    Json::Value configurations = root["configurations"];
    for ( int i = 0; i < configurations.size(); ++i ) {
        StatConfigurations* w = new StatConfigurations();
        w->Deserialize(configurations[i]);
        this->configurations->push_back(w);
    }
    
}

File: dotNet/Semantria/Mapping/Output/Statistics.cs
using System;
using System.Runtime.Serialization;
using System.Xml.Serialization;
using System.Collections.Generic;
using System.ComponentModel;

namespace Semantria.Com.Mapping.Output
{
    [DataContract(Name = "statistics")]
    [XmlRootAttribute("statistics", Namespace = "")]
    [DefaultProperty("Name")]
    public class Statistics : StatisticsBase
    {
        [ReadOnly(true), Category("Common"), DisplayName("Subscriber name"), Description("Subscriber name")]
        [DataMember(Name = "name")]
        [XmlElementAttribute("name")]
        public override string Name { get; set; }

        [ReadOnly(true), Category("Common"), DisplayName("Subscriber status"), Description("Subscriber status")]
        [DataMember(Name = "status")]
        [XmlElementAttribute("status")]
        public string Status { get; set; }

        [Browsable(false)]
        [DataMember(Name = "configurations")]
        [XmlArrayAttribute("configurations")]
        [XmlArrayItemAttribute("configuration", typeof(StatisticsConfiguration))]
        public List<StatisticsConfiguration> Configurations { get; set; }
    }

    [DataContract(Name = "statistic")]
    [XmlRootAttribute("statistic", Namespace = "")]
    [DefaultProperty("Name")]
    public class StatisticsConfiguration : StatisticsBase
    {
        [Browsable(false)]
        [DataMember(Name = "config_id")]
        [XmlElementAttribute("config_id")]
        public string ConfigId { get; set; }
    }

    [DataContract(Name = "statistic")]
    [XmlRootAttribute("statistic", Namespace = "")]
    public abstract class StatisticsBase
    {
        [ReadOnly(true), Category("Common"), DisplayName("Configuration name"), Description("Configuration name")]
        [DataMember(Name = "name")]
        [XmlElementAttribute("name")]
        public virtual string Name { get; set; }

        [ReadOnly(true), Category("Overall"), DisplayName("Processed texts"), Description("Total number of processed texts")]
        [DataMember(Name = "overall_texts")]
        [XmlElementAttribute("overall_texts")]
        public int OverallTexts { get; set; }

        [ReadOnly(true), Category("Overall"), DisplayName("Queued batches"), Description("Number of queued batches")]
        [DataMember(Name = "overall_batches")]
        [XmlElementAttribute("overall_batches")]
        public int OverallBatches { get; set; }

        [ReadOnly(true), Category("Overall"), DisplayName("Overall API calls"), Description("Total number of all performed API calls")]
        [DataMember(Name = "overall_calls")]
        [XmlElementAttribute("overall_calls")]
        public int OverallCalls { get; set; }

		[ReadOnly(true), Category("Overall"), DisplayName("Overall texts exceeded"), Description("Total number of all texts exceeded suggested limit")]
		[DataMember(Name = "overall_exceeded")]
		[XmlElementAttribute("overall_exceeded")]
		public int OverallExceeded { get; set; }

        [ReadOnly(true), Category("Overall"), DisplayName("System API calls"), Description("Number of performed system API calls")]
        [DataMember(Name = "calls_system")]
        [XmlElementAttribute("calls_system")]
        public int CallsSystem { get; set; }

        [ReadOnly(true), Category("Overall"), DisplayName("Data API calls"), Description("Number of performed data calls")]
        [DataMember(Name = "calls_data")]
        [XmlElementAttribute("calls_data")]
        public int CallsData { get; set; }

        [ReadOnly(true), Category("Overall document"), DisplayName("Queued documents"), Description("Total number of queued documents")]
        [DataMember(Name = "overall_docs")]
        [XmlElementAttribute("overall_docs")]
        public int OverallDocs { get; set; }

        [ReadOnly(true), Category("Overall document"), DisplayName("Processed documents"), Description("Number of PROCESSED documents")]
        [DataMember(Name = "docs_processed")]
        [XmlElementAttribute("docs_processed")]
        public int DocsProcessed { get; set; }

        [ReadOnly(true), Category("Overall document"), DisplayName("Failed documents"), Description("Number of FAILED documents ")]
        [DataMember(Name = "docs_failed")]
        [XmlElementAttribute("docs_failed")]
        public int DocsFailed { get; set; }

        [ReadOnly(true), Category("Overall document"), DisplayName("Returned documents"), Description("Number of returned documents")]
        [DataMember(Name = "docs_responded")]
        [XmlElementAttribute("docs_responded")]
        public int DocsResponded { get; set; }

        [ReadOnly(true), Category("Overall collection"), DisplayName("Queued collections"), Description("Total number of queued collections")]
        [DataMember(Name = "overall_colls")]
        [XmlElementAttribute("overall_colls")]
        public int OverallColls { get; set; }

        [ReadOnly(true), Category("Overall collection"), DisplayName("Processed collections"), Description("Number of PROCESSED collections")]
        [DataMember(Name = "colls_processed")]
        [XmlElementAttribute("colls_processed")]
        public int CollsProcessed { get; set; }

        [ReadOnly(true), Category("Overall collection"), DisplayName("Failed collections"), Description("Number of FAILED collections")]
        [DataMember(Name = "colls_failed")]
        [XmlElementAttribute("colls_failed")]
        public int CollsFailed { get; set; }

        [ReadOnly(true), Category("Overall collection"), DisplayName("Returned collections"), Description("Number of returned collections")]
        [DataMember(Name = "colls_responded")]
        [XmlElementAttribute("colls_responded")]
        public int CollsResponded { get; set; }

        [ReadOnly(true), Category("Overall collection"), DisplayName("Documents in collections"), Description("Total number of documents queued within collections")]
        [DataMember(Name = "colls_documents")]
        [XmlElementAttribute("colls_documents")]
        public int CollsDocuments { get; set; }

		[ReadOnly(true), Category("Common"), DisplayName("Last used app"), Description("The latest used application")]
		[DataMember(Name = "latest_used_app")]
		[XmlElementAttribute("latest_used_app")]
		public string LatestUsedApp { get; set; }

		[ReadOnly(true), Category("Common"), DisplayName("All used apps"), Description("All used applications")]
		[DataMember(Name = "used_apps")]
		[XmlElementAttribute("used_apps")]
		public string UsedApp { get; set; }
    }
}
